File: content/snippets/scala2/interview_lcci_16_10_living_people.scala
object Lcci1610LivingPeople {
    def maxAliveYear(birth: Array[Int], death: Array[Int]): Int = {
        val diff = Array.fill(102)(0)
        for (i <- birth.indices) {
            diff(birth(i) - 1900) += 1
            if (death(i) + 1 <= 2000) {
                diff(death(i) + 1 - 1900) -= 1
            }
        }

        var bestYear = 1900
        var alive = 0
        var bestAlive = -1
        for (i <- 0 to 100) {
            alive += diff(i)
            if (alive > bestAlive) {
                bestAlive = alive
                bestYear = 1900 + i
            }
        }
        bestYear
    }

    def main(args: Array[String]): Unit = {
        println(maxAliveYear(Array(1900, 1901, 1950), Array(1948, 1951, 2000)))
    }
}

File: content/snippets/cpp/interview_lcci_16_10_living_people.cpp
#include <iostream>
#include <vector>
using namespace std;

int maxAliveYear(const vector<int>& birth, const vector<int>& death) {
    vector<int> diff(102, 0);
    for (int i = 0; i < static_cast<int>(birth.size()); ++i) {
        ++diff[birth[i] - 1900];
        if (death[i] + 1 <= 2000) {
            --diff[death[i] + 1 - 1900];
        }
    }
    int bestYear = 1900;
    int alive = 0;
    int bestAlive = -1;
    for (int i = 0; i <= 100; ++i) {
        alive += diff[i];
        if (alive > bestAlive) {
            bestAlive = alive;
            bestYear = 1900 + i;
        }
    }
    return bestYear;
}

int main() {
    cout << maxAliveYear({1900, 1901, 1950}, {1948, 1951, 2000}) << endl;
    return 0;
}
